### DISCUSSION TITLE
ft2build.h on downloading devtools

### DISCUSSION THREAD

**Discussion by yaaman3** [nig-sc/nigsc_homepage2#324] [Q&A] [answered]
R devtoolsのダウンロードについてご相談します。
遺伝研のウェブサイト記載の方法に従って、spackを使用して、r-4.2.2を使用しています。
https://sc.ddbj.nig.ac.jp/software/R

[> > install.packages("devtools")
> (途中省略)
> ```
> * installing *source* package ‘systemfonts’ ...
> ** package ‘systemfonts’ successfully unpacked and MD5 sums checked
> ** using staged installation
> Package fontconfig was not found in the pkg-config search path.
> Perhaps you should add the directory containing `fontconfig.pc'
> to the PKG_CONFIG_PATH environment variable
> Package 'fontconfig', required by 'virtual:world', not found
> Using PKG_CFLAGS=
> Using PKG_LIBS=-lfontconfig -lfreetype
> ** libs
> rm -f systemfonts.so caches.o cpp11.o dev_metrics.o font_matching.o font_registry.o ft_cache.o string_shape.o font_metrics.o font_fallback.o string_metrics.o emoji.o cache_store.o init.o unix/FontManagerLinux.o
> /lustre7/home/********/spack/opt/spack/linux-centos7-zen2/gcc-9.2.0/gcc-9.2.0-osduzt533fkkujuih7xd6q7cnv63hvhu/bin/g++ -std=gnu++11 -I"/lustre7/home/nakayan20/spack/opt/spack/linux-centos7-zen2/gcc-9.2.0/r-4.2.2-zqzktwlaym6ti4o3lmxokitst5mcpvnz/rlib/R/include" -DNDEBUG  -I'/lustre7/home/********/spack/opt/spack/linux-centos7-zen2/gcc-9.2.0/r-4.2.2-zqzktwlaym6ti4o3lmxokitst5mcpvnz/rlib/R/library/cpp11/include' -I/usr/local/include   -fpic  -g -O2  -c caches.cpp -o caches.o
> In file included from caches.h:7,
>                  from caches.cpp:1:
> ft_cache.h:9:10: fatal error: ft2build.h: No such file or directory
>     9 | #include <ft2build.h>
>       |          ^~~~~~~~~~~~
> compilation terminated.
> ```](url)

devtoolsダウンロード時に、依存対象のsystemfotsでエラーが出ます。
このエラーを検索すると、ft2build.hは、freetype-develに含まれており、libfreetype6-devを導入すれば対応可能とのことです。
sudo権限なしで、スパコン上で、こちらのパッケージは導入可能でしょうか。
初歩的なことで恐縮ですが、ご教授ください。

**Comment by oogasawa** [comment]
こちらでまだ動作確認していませんが、
spackのpackage listのページを見ると
https://spack.readthedocs.io/en/latest/package_list.html
r-devtoolsというspackパッケージがあるので、
https://spack.readthedocs.io/en/latest/package_list.html#r-devtools
比較的簡単に解決できる可能性のある一つの方法はRからinstall.packages("devtools")する代わりに、
これをspack install r-devtoolsすればdevtoolsが入った状態のRがインストールされるのではないかと思います。

**Reply by oogasawa** [reply]
これも完全には動作確認しておらずすみませんが、RをspackでインストールするときにはX=Trueのオプションを付けないと描画関係でいろいろインストールされていない旨のエラーが頻発するかもしれません。一方X=Trueを付けるとコンパイル時間が非常に長くなるので -j 8等コンパイル時の並列度を上げて実行する必要があると思います。

```
spack install r X=True
```

**Reply by yaaman3** [reply]
返信をくださり、ありがとうございます。実行してみました。
1点目の r-devtools：`$ spack install r-devtools`
2点目の X=True付加：`$ spack install r@4.2.1 X=True`
libxcb, pythonで共通するエラーが生じて、ダウンロードできませんでした。

libxcbのエラー (途中省略)
```
[+] /lustre7/home/**********/spack/opt/spack/linux-centos7-zen2/gcc-9.2.0/libsm-1.2.3-27haj7t5zbeox5d5qiscmsuycuvryr6c
==> Installing libxcb-1.14-m5yxzjpxx46rt3ppgfjeh5tlijpngmrt
==> No binary for libxcb-1.14-m5yxzjpxx46rt3ppgfjeh5tlijpngmrt found: installing from source
==> Using cached archive: /lustre7/home/*********/spack/var/spack/cache/_source-cache/archive/a5/a55ed6db98d43469801262d81dc2572ed124edc3db31059d4e9916eb9f844c34.tar.xz
==> Ran patch() for libxcb
==> libxcb: Executing phase: 'autoreconf'
==> libxcb: Executing phase: 'configure'
==> libxcb: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

1 error found in build log:
     5703    /tmp/*********/spack-stage/spack-stage-libxcb-1.14-m5yxzjpxx46rt3ppgfjeh5tlijpngmrt/spack-src/src/xvmc.h:471: warning:
             The following parameters of xcb_xvmc_list_surface_types(xcb_connection_t *c, xcb_xv_port_t port_id) are not documented:
     5704      parameter 'port_id'
     5705    /tmp/*********/spack-stage/spack-stage-libxcb-1.14-m5yxzjpxx46rt3ppgfjeh5tlijpngmrt/spack-src/src/xvmc.h:483: warning:
             The following parameters of xcb_xvmc_list_surface_types_unchecked(xcb_connection_t *c, xcb_xv_port_t port_id) are not d
             ocumented:
     5706      parameter 'port_id'
     5707    make[1]: *** [manual] Segmentation fault (core dumped)
     5708    make[1]: Leaving directory `/tmp/**********/spack-stage/spack-stage-libxcb-1.14-m5yxzjpxx46rt3ppgfjeh5tlijpngmrt/spack-s
             rc/doc'
  >> 5709    make: *** [all-recursive] Error 1
```
確信を持てませんが、Spack公式サイトのknown issues "Variants are not properly forwarded to dependencies" という項目と関連があるかと思いました。
https://spack.readthedocs.io/en/v0.13.2/known_issues.html

**Comment by sc-fs** [comment]
既存のspackの環境にて
`$ spack install r-devtools `
を実行した場合、はじめに指摘のあったエラーft2build.h: No such file or directoryが発生しました。
環境が違う可能性がありましたので、
`$ spack uninstall --all `でクリアを実行し、
コンパイラ(gcc@9.2.0)から入れ直して、無事インストールできました。
```
$ spack install r-devtools
$ spack load r-devtools
$ R
  R version 4.2.2 (2022-10-31) -- "Innocent and Trusting"
  
  （省略）

> library(devtools)
Loading required package: usethis

Attaching package: ‘devtools’

The following object is masked from ‘package:testthat’:

    test_file

>

```
一度　spackの環境をクリアしてから、再度インストールしていただけますか。

**Reply by yaaman3** [reply, marked as answer]
ご指示いただいた方法で、無事にインストールできました。
この度はありがとうございました。
